File: trees/lower_common_ancestor_bst.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# Approach 1: Recursive
# @param {TreeNode} root
# @param {TreeNode} p
# @param {TreeNode} q
# @return {TreeNode}
def lowest_common_ancestor(root, p, q)
    return root if root.nil? || root == p || root == q

    return lowest_common_ancestor(root.left, p, q) if p.val < root.val && q.val < root.val
    return lowest_common_ancestor(root.right, p, q) if p.val > root.val && q.val > root.val

    root
end

# Approach 2: Iterative
def lowest_common_ancestor(root, p, q)
    while root
        if p.val < root.val && q.val < root.val
            root = root.left
        elsif root.val < p.val && root.val < q.val
            root = root.right
        else
            return root
        end
    end
end


# Approach 1: Recursive
# Approach 2: Iterative

File: trees/lower_common_ancestor_bst.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

# Approach 1: Recursive
class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        return self.dfs(root, p, q)

    def dfs(self, root, p, q):
        if not root or root == p or root == q: return root

        if p.val < root.val and q.val < root.val: return self.dfs(root.left, p, q)
        if p.val > root.val and q.val > root.val: return self.dfs(root.right, p, q)

        return root

# Approach 2: Iterative
class Solution2:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        while root:
            if p.val < root.val and q.val < root.val:
                root = root.left
            elif root.val < p.val and root.val < q.val:
                root = root.right
            else:
                return root


# 235. Lowest Common Ancestor of a Binary Search Tree
# https://leetcode.com/problems/lowest-common-ancestor-of-a-binary-search-tree/description/

# Approach 1: Recursive
# Approach 2: Iterative
